File: 백준/Gold/7576. 토마토/토마토.py
import sys
from collections import deque

m,n = map(int,sys.stdin.readline().split())
arr = []
max = 1
cnt = 0
for i in range(n):
    arr.append(list(map(int,sys.stdin.readline().split())))
dq = deque()
for i in range(n):
    for j in range(m):
        if arr[i][j] == 1:
            dq.append([i,j])
        if arr[i][j] == -1:
            cnt += 1

while dq:
        tmp = dq.popleft()
        cnt +=1
        if tmp[0]+1 < n and arr[tmp[0]+1][tmp[1]] == 0: # 아래
            arr[tmp[0]+1][tmp[1]] = arr[tmp[0]][tmp[1]] + 1
            dq.append([tmp[0]+1,tmp[1]])
            if max < arr[tmp[0]][tmp[1]] + 1:
                max = arr[tmp[0]][tmp[1]] + 1
        if 0 <= tmp[0]-1 and arr[tmp[0]-1][tmp[1]] == 0: # 위
            arr[tmp[0]-1][tmp[1]] = arr[tmp[0]][tmp[1]] + 1
            dq.append([tmp[0]-1,tmp[1]])
            if max < arr[tmp[0]][tmp[1]] + 1:
                max = arr[tmp[0]][tmp[1]] + 1
        if tmp[1]+1 < m  and arr[tmp[0]][tmp[1]+1] == 0: # 오른쪽
            arr[tmp[0]][tmp[1]+1] = arr[tmp[0]][tmp[1]] + 1
            dq.append([tmp[0],tmp[1]+1])
            if max < arr[tmp[0]][tmp[1]] + 1:
                max = arr[tmp[0]][tmp[1]] + 1
        if 0 <= tmp[1]-1 and arr[tmp[0]][tmp[1]-1] == 0: # 왼쪽
            arr[tmp[0]][tmp[1]-1] = arr[tmp[0]][tmp[1]] + 1
            dq.append([tmp[0],tmp[1]-1])
            if max < arr[tmp[0]][tmp[1]] + 1:
                max = arr[tmp[0]][tmp[1]] + 1
if n*m - cnt == 0:
    print(max-1)
else:
    print(-1)

File: 백준/Gold/7576. 토마토/토마토.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int m = sc.nextInt();
        int n = sc.nextInt();

        int[][] arr = new int[n][m];
        int max = 1;
        int cnt = 0;

        Queue<int[]> dq = new LinkedList<>();

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                arr[i][j] = sc.nextInt();
                if (arr[i][j] == 1) {
                    dq.offer(new int[]{i, j});
                }
                if (arr[i][j] == -1) {
                    cnt++;
                }
            }
        }

        int[][] move = {{1,0}, {-1,0}, {0,1}, {0,-1}};

        while (!dq.isEmpty()) {
            int[] tmp = dq.poll();
            cnt++;

            for (int[] mv : move) {
                int nx = tmp[0] + mv[0];
                int ny = tmp[1] + mv[1];

                if (nx >= 0 && nx < n && ny >= 0 && ny < m && arr[nx][ny] == 0) {
                    arr[nx][ny] = arr[tmp[0]][tmp[1]] + 1;
                    dq.offer(new int[]{nx, ny});
                    max = Math.max(max, arr[nx][ny]);
                }
            }
        }

        if (n * m - cnt == 0) {
            System.out.println(max - 1);
        } else {
            System.out.println(-1);
        }

        sc.close();
    }
}